src/plugins/discourse/mirror.test.js

```diff
@@ -132,6 +164,20 @@ class MockFetcher implements Discourse {
     };
   }
 
+  addCategory(topicId: TopicId): PostId {
```

Comment:
What does it mean to "add" a category? And why does it return a postId? I am quite confused.

Reply:
Hah, right. I wrote up a reply of how that happened. The gist being it builds on the previous MockFetcher design where Topics were created almost entirely by convention. But this just makes it worse :sweat_smile:.

How about, I change the below function to be:
`setTopicCategory(topicId: TopicId, categoryId: CategoryId | null, isDefinition: boolean)`

So you can do:
```js
// Create Topic ID 4, with a post.
const p1 = addPost(4, null);
// Set Topic ID 4 as the category definition topic for Category ID 6.
setTopicCategory(4, 6, true);

// Create Topic ID 5 with a post.
const p2 = addPost(5, null);
// Set Topic ID 5 as a normal topic in Category ID 6.
setTopicCategory(4, 6, false);
```

Reply:
It is confusing. It comes from MockFetcher generating Topics implicitly, by calling `addPost` with a new TopicId.
This means, TopicIds are planned, but created through Posts. PostIds are incremental and returned by `addPost`.

The change here introduces new concepts. Categories and category definition topics.
Categories, like Topics are also generated implicitly.
And Topics are:
- Uncategorized when not in the `_topicCategories` mapping.
- Category definition topics when, TopicId == CategoryId.
- Normal topics in a category when, TopicId != CategoryId.

The way to create a topic is by adding it's first post.
```js
// Creates topic: 4, with post: p1.
const p1 = addPost(4, null);
```

The way to have a topic in a category normally:
```js
// Adds that topic to category ID 6.
setTopicCategory(4, 6);
```

But a category should have a category definition topic.
One way to do so, using just the convention of "Category definition topics when, TopicId == CategoryId"
```js
// Creates topic: 6, with post: p2.
const p2 = addPost(6, null);
// Adds that topic to category ID 6, which makes it a category definition topic.
setTopicCategory(6, 6);
```

The functional equivalent of that is:
```js
// Creates topic: 6, with post: p2.
// And adds the topic to category ID 6, which makes it a category definition topic.
const p2 = addCategory(6);
```

---

But that certainly doesn't make anything any clearer.
Perhaps the process should be inverted. So the signature is:
`defineCategory(topicId: TopicId): CategoryId`

Throwing if it's called with a topicId that doesn't exist in the MockFetcher yet.
Doing away with the TopicId == CategoryId convention.

It would look something like:
```js
// Create topic: 5, with post: p1.
const p1 = addPost(5, null);
// Create category: c1, with definition topic: 5.
const c1 = defineCategory(5);

// Create topic: 7, with post: p2.
const p2 = addPost(7, null);
// Add it to category: c1.
setTopicCategory(7, c1);
```